deb/deb.go

```diff
@@ -80,27 +81,60 @@ func (*Deb) Package(info *nfpm.Info, deb io.Writer) (err error) {
 	if ok {
 		info.Arch = arch
 	}
+
 	dataTarGz, md5sums, instSize, err := createDataTarGz(info)
 	if err != nil {
 		return err
 	}
+
 	controlTarGz, err := createControl(instSize, md5sums, info)
 	if err != nil {
 		return err
 	}
+
+	debianBinary := []byte("2.0\n")
+
 	var w = ar.NewWriter(deb)
 	if err := w.WriteGlobalHeader(); err != nil {
 		return errors.Wrap(err, "cannot write ar header to deb file")
 	}
-	if err := addArFile(w, "debian-binary", []byte("2.0\n")); err != nil {
+
+	if err := addArFile(w, "debian-binary", debianBinary); err != nil {
 		return errors.Wrap(err, "cannot pack debian-binary")
 	}
+
 	if err := addArFile(w, "control.tar.gz", controlTarGz); err != nil {
 		return errors.Wrap(err, "cannot add control.tar.gz to deb")
 	}
+
 	if err := addArFile(w, "data.tar.gz", dataTarGz); err != nil {
 		return errors.Wrap(err, "cannot add data.tar.gz to deb")
 	}
+
+	if info.Deb.Signature.KeyFile != "" {
+		data := io.MultiReader(bytes.NewReader(debianBinary), bytes.NewReader(controlTarGz),
+			bytes.NewReader(dataTarGz))
+
+		sig, err := sign.PGPArmoredDetachSign(data, info.Deb.Signature.KeyFile,
+			info.Deb.Signature.KeyPassphrase)
+		if err != nil {
+			return errors.Wrap(err, "create signature")
```

Comment:
I would put all these errors as exported errors, so API users can check against them...

Reply:
Also you do not need to use `errors.Wrap` any more as that is built into go 1.13 where you can use `%w` https://blog.golang.org/go1.13-errors

Though it might be best to leave as is for now and do an explicit PR for removing `errors.Wrap`

Reply:
> I would put all these errors as exported errors, so API users can check against them...

Where would I export them? This is an internal package that should not be imported. Also, I would still wrap the underlying errors in the newly exported ones.

> Also you do not need to use errors.Wrap any more as that is built into go 1.13 where you can use %w https://blog.golang.org/go1.13-errors
>
> Though it might be best to leave as is for now and do an explicit PR for removing errors.Wrap

I agree, we should change wrapping over the whole codebase at some point. When I started working on `nfpm` `golangci-lint` would not allow me to use `%w`, but I just tried it again and now it works. Maybe I was using an outdated `golangci-lint` version from `brew` before or something.

Anyway, I'm a big proponent of `%w` and I already combed through some large codebases of mine to replace every `%v` in errors with `%w` 😄

Reply:
> Where would I export them? This is an internal package that should not be imported. Also, I would still wrap the underlying errors in the newly exported ones.

`deb` is not an internal package... I would export them on the deb.go file 🤔

----

agree, lets migrate from errors.wrap in the near future 🙏

Reply:
> deb is not an internal package... I would export them on the deb.go file 🤔

You are right, but the other errors in your other comments are from an internal package

Reply:
maybe just one, and wrap them? Maybe even on the main package (e.g. `nfpm.SigningError`) and then wrap it here 🤔

Reply:
Do I understand you correctly that I should wrap the primitive errors thrown by `internal/sign` in an `nfpm.SigningError` and return that?

Reply:
yeah, that's what I meant... but tbh, maybe in another PR? not sure if this will ever be needed... 🤔

Reply:
Like this?

```go
type SigningError struct {
	error
}

func (s *SigningError) String() string {
	return fmt.Sprintf("signing error: %v", s.error)
}

func (s *SigningError) Unwarp() error {
	return s.error
}

func (s *SigningError) Is(target error) bool {
	_, ok := target.(*SigningError)
	return ok
}```

Reply:
Hmm, I think we should really only add it as soon as someone needs this...

Reply:
I would say just:

```go
var ErrSigning = errors.New("signing errors")
```

and then we can wrap errors as we already do (and later migrate to the new go way) and everything should work.

also your `Is` implementation is not correct I think... it should check for error equality, not type (yeah, its confusing). For type check we should use `errors.As` 🤷‍♂️

Reply:
But that would mean discarding the underlying error which is very useful to figure out what actually went wrong. Or did I miss something?

Reply:
using the `fmt.Errorf("something: %w")` you can call `errors.Unwrap` to get the `SigningError` and also check type with `errors.As` (or, since this a "simple" error (not a `struct` that implements the error interface), `errors.Is` should work as well with no additional changes).

so, tl;dr, the caller could do:

```go
if errors.Is(err, nfpm.SigningError) {
```

and everything should work as expected. 🤔

Reply:
Sure but if `ErrSigning` is constructed with `errors.New` it will always be a "leaf" error, that cannot be unwrapped further. So everything that goes wrong in the signing process will only be reported as the general `ErrSigning`. Then the users won't know if the key file name is misspelled or the password is unset or wrong.

Reply:
This is why I wanted to implement `ErrSinging` in a way that it still holds the unwrapable underlying error thrown by the `internal/sign` package.

Reply:
hmm, yeah, that makes sense... 

let's implement your idea then.

I think you don't need the `Is` method though, only `Unwrap` :)

Reply:
(and `Error` of course)

Reply:
Yep, I'll do it tomorrow.